### PR TITLE
test: Fix flaky e2e tests | Add retry logic to switchToWindowWithTitle | Update retry helper fn

### LINKED ISSUE #21416: e2e fix flaky No window with title: <Window Title> tests
## **Description**
We've recently had a [spike in flaky e2e tests](https://app.circleci.com/insights/github/MetaMask/metamask-extension/workflows/test_and_release/tests). We observed these tests to have a common denominator: `switchToWindowWithTitle`.

"No window with title: <Window Title>" e.g. No window with title: MetaMask Notifications" caused flaky tests.

Links:
- [CircleCI Insights - Flaky Tests](https://app.circleci.com/insights/github/MetaMask/metamask-extension/workflows/test_and_release/tests)
- [Internal discussion](https://consensys.slack.com/archives/GTQAGKY5V/p1697150920731179)

## **Manual testing steps**

1. Run CI/CD to observe tests

### **Before**

Screenshot 2023-10-16 21:00:

Screenshot 2023-10-16 21:50:

### PR DESCRIPTION
**Pull request by digiwand**
## **Description**
We've recently had a [spike in flaky e2e tests](https://app.circleci.com/insights/github/MetaMask/metamask-extension/workflows/test_and_release/tests). We observed these tests to have a common denominator: `switchToWindowWithTitle`. This PR adds retry logic to `switchToWindowWithTitle`. 

After the first try,  the logic will retry a max number of 8 times with 2500ms delay in-between. These numbers are arbitrary and generous just in case.

This PR also updates the retry logic in `development/lib/retry.js` to return the function value on success. 

Links:
- [CircleCI Insights - Flaky Tests](https://app.circleci.com/insights/github/MetaMask/metamask-extension/workflows/test_and_release/tests)
- [Internal discussion](https://consensys.slack.com/archives/GTQAGKY5V/p1697150920731179)

## **Manual testing steps**

1. Run CI/CD to observe tests

### **Before**

Screenshot 2023-10-16 21:00:

Screenshot 2023-10-16 21:50:

## **Related issues**

Fixes https://github.com/MetaMask/metamask-extension/issues/21416

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained:
  - [ ] What problem this PR is solving.
  - [ ] How this problem was solved.
  - [ ] How reviewers can test my changes.
- [ ] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by HowardBraham** [comment]
I don't think this change is going to be effective.  Because it's in a `for` loop, it's effectively already retrying.

@k-g-j (a brand new MetaMask team member) and I are also working on this problem.

**Comment by digiwand** [comment]
@HowardBraham 

hm... if it's the for loop I'm thinking of `for (const handle of windowHandles) {`, I think this only tries to find each window once with a delay in-between. It does not appear to catch any errors or retry logic.

I thought I saw in a failed test where it could not find the popup window by title "MetaMask Notifications" which led me to propose this solution

**Comment by digiwand** [comment]
so far a good sign that in ran successfully on the first try. Updating constraints and running again

**Comment by HowardBraham** [comment]
Sorry, you're right, it's the outer loop
```
while (timeElapsed <= timeout) {
  for (const handle of windowHandles) {
```

It will try swtiching `timeout / delayStep` times.

We're going to add some better tests for when new windows open and close.

**Comment by digiwand** [comment]
yeah which is also interesting. timeElapsed was 5 seconds and was [increased to 10 seconds](https://github.com/MetaMask/metamask-extension/pull/19708) a few months ago. Maybe with this logic we could lower timeElapsed back down.   

sounds good. not planning to add anything extra to this PR. We could make a follow-up PR to fine tune the methods as you alluded to in our [internal thread](https://consensys.slack.com/archives/GTQAGKY5V/p1697223609357109?thread_ts=1697150920.731179&cid=GTQAGKY5V). I think it's a good idea

**Comment by digiwand** [comment]
once ci/circleci: test-deps-audit tests are fixed, this PR can be ready for review

test dep @babel/traverse PR:
https://github.com/MetaMask/metamask-extension/pull/21393

**Review by legobeat — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]

File: development/lib/retry.js [hunk 1]
```diff
@@ -14,8 +14,10 @@
  * @param {string} args.retryUntilFailure - Retries until the function fails.
  * @param {Function} functionToRetry - The function that is run and tested for
  * failure.
- * @returns {Promise<null | Error>} a promise that either resolves to null if
- * the function is successful or is rejected with rejectionMessage otherwise.
+ * @returns {Promise<* | null | Error>} a promise that either resolves with one of the following:
```
[the remark this answers is not in the document]
Comment (digiwand) [reply development/lib/retry.js:17]: Agreed. This method and surrounding methods could use improvements.

p.s. I don't have bandwidth for at least the next 1.5 weeks. quest to do in this time. 

created https://github.com/MetaMask/metamask-extension/issues/21403 to document requested improvements

**Pull request merged**